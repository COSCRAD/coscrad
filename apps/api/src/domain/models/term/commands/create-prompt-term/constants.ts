export const CREATE_PROMPT_TERM = `CREATE_PROMPT_TERM`;

export const PROMPT_TERM_CREATED = `PROMPT_TERM_CREATED`;
