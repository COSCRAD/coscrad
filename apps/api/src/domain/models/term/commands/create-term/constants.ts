export const CREATE_TERM = `CREATE_TERM`;

export const TERM_CREATED = `TERM_CREATED`;
