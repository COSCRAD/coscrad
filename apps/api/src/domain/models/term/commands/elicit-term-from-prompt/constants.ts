export const ELICIT_TERM_FROM_PROMPT = `ELICIT_TERM_FROM_PROMPT`;

export const PROMPT_FROM_ELICIT_TERM = `PROMPT_FROM_ELICIT_TERM`;
