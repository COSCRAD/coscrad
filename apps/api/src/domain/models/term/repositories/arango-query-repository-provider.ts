import { DiscoveryService } from '@golevelup/nestjs-discovery';
import { InternalError } from '../../../../lib/errors/InternalError';
import { isNotFound } from '../../../../lib/types/not-found';
import {
    getArangoViewRepositoryMetadata,
    hasArangoViewRepositoryMetatdata,
} from '../../../../persistence/database/decorators/arango-view-repository.decorator';
import {
    IPublishable,
    IQueryRepositoryProvider,
} from '../../shared/common-commands/publish-resource/resource-published.event-handler';

/**
 * TODO We need to find a pattern to make this more extensible. Maybe we should
 * have a decorator that declares the query repository.
 * @CoscradQueryRepository(ResourceType.term).
 * Then we can use reflection to return the desired repository.
 */
export class ArangoQueryRepositoryProvider implements IQueryRepositoryProvider {
    private readonly viewTypeToRepository = new Map<string, unknown>();

    private isInitialized = false;

    //    TODO Use the standard []`Reflector` pattern](https://docs.nestjs.com/fundamentals/execution-context#reflection-and-metadata) from `NestJS`
    constructor(private readonly discoveryService: DiscoveryService) {}

    // can this be `forView(viewtype: string)` ?
    forView<T extends IPublishable>(viewType: string): T {
        if (!this.viewTypeToRepository.has(viewType)) {
            throw new InternalError(
                `Cannot provide a query repository for unknown view type: ${viewType}`
            );
        }

        return this.viewTypeToRepository.get(viewType) as T;
    }

    public async initialize() {
        if (this.isInitialized) {
            return;
        }

        const queryRepositoryProviders = await this.discoveryService.providers((provider) =>
            hasArangoViewRepositoryMetatdata(provider?.injectType)
        );

        for (const provider of queryRepositoryProviders) {
            const meta = getArangoViewRepositoryMetadata(provider.injectType);

            if (isNotFound(meta)) return;

            this.register(meta.viewType, provider.instance);
        }

        this.isInitialized = true;
    }

    private register(viewType: string, repository: unknown) {
        if (!this.viewTypeToRepository.has(viewType)) {
            this.viewTypeToRepository.set(viewType, repository);
        }
    }
}
