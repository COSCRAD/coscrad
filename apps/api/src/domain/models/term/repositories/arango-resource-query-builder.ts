import { LanguageCode, MultilingualTextItemRole } from '@coscrad/api-interfaces';
import { AqlQuery } from 'arangojs/aql';
import { AggregateId } from '../../../types/AggregateId';
import { BaseEvent } from '../../shared/events/base-event.entity';
import { CoscradDate } from '../../user-management/utilities';

export class ArangoResourceQueryBuilder {
    constructor(private readonly collectionName: string) {}

    publish(id: AggregateId): AqlQuery {
        const query = `
        FOR doc IN @@collectionName
        FILTER doc._key == @id
        UPDATE doc WITH {
            isPublished: true,
            actions: REMOVE_VALUE(doc.actions,"PUBLISH_RESOURCE")
        } IN @@collectionName
        `;

        const bindVars = {
            '@collectionName': this.collectionName,
            id,
        };

        return {
            query,
            bindVars,
        };
    }

    allowUser(resourceId: AggregateId, userId: AggregateId): AqlQuery {
        const query = `
                FOR doc IN @@collectionName
                FILTER doc._key == @id
                UPDATE doc WITH {
                    accessControlList: {
                        allowedUserIds: APPEND(doc.accessControlList.allowedUserIds,[@userId])
                    }
                } IN @@collectionName
                 RETURN OLD
                `;
        // TODO remove return value?

        const bindVars = {
            '@collectionName': this.collectionName,
            id: resourceId,
            userId,
        };

        return {
            query,
            bindVars,
        };
    }

    translateName(
        id: AggregateId,
        text: String,
        languageCode: LanguageCode,
        role: MultilingualTextItemRole
    ): AqlQuery {
        const query = `
            FOR doc IN @@collectionName
            FILTER doc._key == @id
            let newItem = {
                        text: @text,
                        languageCode: @languageCode,
                        role: @role
            }
            UPDATE doc WITH {
                name: {
                    items: APPEND(doc.name.items,newItem)
                }
            } IN @@collectionName
            `;

        const bindVars = {
            '@collectionName': this.collectionName,
            id: id,
            text: text,
            // TODO we may want this to be passed in, presumably from an event payload
            role,
            languageCode: languageCode,
        };

        return {
            query,
            bindVars,
        };
    }

    attribute(resourceId: AggregateId, event: BaseEvent): AqlQuery {
        const query = `
                FOR doc IN @@collectionName
                FILTER doc._key == @id
                LET contributorsForThisEvent = (
                    FOR contributorId IN @contributorIds
                    FOR c IN contributors
                    FILTER c._key == contributorId
                    RETURN c
                )
                LET listOfContributors = (
                    for c in contributorsForThisEvent
                    return CONCAT_SEPARATOR(' ',[c.fullName.firstName,c.fullName.lastName])
                )
                LET contributorIds = (
                    for c in contributorsForThisEvent
                    return c._key
                )
                LET attribution = CONCAT(@template,LENGTH(listOfContributors)>0 ? CONCAT_SEPARATOR(', ',listOfContributors) : "admin")
                LET newContributions = {
                    type: @eventType,
                    contributorIds,
                    statement: attribution,
                    date: @date,
                    timestamp: @timestamp
                }
                LET updatedContributions = APPEND(doc.contributions,newContributions)
                UPDATE doc WITH {
                    contributions: updatedContributions
                } IN @@collectionName
                 RETURN updatedContributions
                `;

        const bindVars = {
            '@collectionName': this.collectionName,
            id: resourceId,
            contributorIds: event.meta.contributorIds || [],
            template: event.buildAttributionStatement(),
            date: CoscradDate.fromUnixTimestamp(event.meta.dateCreated),
            timestamp: event.meta.dateCreated,
            eventType: event.type,
        };

        return {
            query,
            bindVars,
        };
    }
}
