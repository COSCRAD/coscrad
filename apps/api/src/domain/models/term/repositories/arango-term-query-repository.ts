import {
    ICommandFormAndLabels,
    IDetailQueryResult,
    IMultilingualTextItem,
    ITermViewModel,
    LanguageCode,
} from '@coscrad/api-interfaces';
import { isNullOrUndefined } from '@coscrad/validation-constraints';
import { Inject } from '@nestjs/common';
import { InternalError } from '../../../../lib/errors/InternalError';
import { Maybe } from '../../../../lib/types/maybe';
import { isNotFound } from '../../../../lib/types/not-found';
import { ArangoConnectionProvider } from '../../../../persistence/database/arango-connection.provider';
import { ArangoDatabase } from '../../../../persistence/database/arango-database';
import { ArangoDatabaseForCollection } from '../../../../persistence/database/arango-database-for-collection';
import mapDatabaseDocumentToAggregateDTO from '../../../../persistence/database/utilities/mapDatabaseDocumentToAggregateDTO';
import mapEntityDTOToDatabaseDocument from '../../../../persistence/database/utilities/mapEntityDTOToDatabaseDocument';
import { AggregateId } from '../../../types/AggregateId';
import {
    AUDIO_QUERY_REPOSITORY_TOKEN,
    IAudioItemQueryRepository,
} from '../../audio-visual/audio-item/queries/audio-item-query-repository.interface';
import { ITermQueryRepository } from '../queries';

export class ArangoTermQueryRepository implements ITermQueryRepository {
    private readonly database: ArangoDatabaseForCollection<
        IDetailQueryResult<ITermViewModel & { actions: ICommandFormAndLabels[] }>
    >;

    constructor(
        arangoConnectionProvider: ArangoConnectionProvider,
        // AUDIO_ITEM_QUERY_REPOSITORY?
        @Inject(AUDIO_QUERY_REPOSITORY_TOKEN)
        private readonly audioItemQueryRepository: IAudioItemQueryRepository
    ) {
        this.database = new ArangoDatabaseForCollection(
            new ArangoDatabase(arangoConnectionProvider.getConnection()),
            'term__VIEWS'
        );
    }

    async create(view: ITermViewModel & { actions: ICommandFormAndLabels[] }): Promise<void> {
        return this.database.create(mapEntityDTOToDatabaseDocument(view));
    }

    async createMany(
        views: (ITermViewModel & { actions: ICommandFormAndLabels[] })[]
    ): Promise<void> {
        return this.database.createMany(views.map(mapEntityDTOToDatabaseDocument));
    }

    async delete(id: AggregateId): Promise<void> {
        return this.database.delete(id);
    }

    async translate(
        id: AggregateId,
        { text, languageCode, role }: IMultilingualTextItem
    ): Promise<void> {
        const query = `
        FOR doc IN @@collectionName
        FILTER doc._key == @id
        let newItem = {
                    text: @text,
                    languageCode: @languageCode,
                    role: @role
        }
        UPDATE doc WITH {
            name: {
                items: APPEND(doc.name.items,newItem)
            }
        } IN @@collectionName
         RETURN OLD
        `;

        const bindVars = {
            '@collectionName': 'term__VIEWS',
            id: id,
            text: text,
            role: role,
            languageCode: languageCode,
        };

        const cursor = await this.database
            .query({
                query,
                bindVars,
            })
            .catch((reason) => {
                throw new InternalError(`Failed to translate term via TermRepository: ${reason}`);
            });

        await cursor.all();
    }

    async addAudio(termId: AggregateId, _languageCode: LanguageCode, audioItemId: string) {
        /**
         * TODO Include this in a single query below to ensure this operation
         * is transactional.
         */
        const audioItemSearchResult = await this.audioItemQueryRepository.fetchById(audioItemId);

        if (isNotFound(audioItemSearchResult)) {
            // TODO log error but still fail gracefully
            return;
        }

        const { mediaItemId, mimeType } = audioItemSearchResult;

        if (isNullOrUndefined(mediaItemId) || isNullOrUndefined(mimeType)) {
            // TODO log error but fail gracefully
            return;
        }

        // note the casing here on `audioURL`
        const query = `
        FOR doc IN @@collectionName
        FILTER doc._key == @id
        UPDATE doc WITH {
            mediaItemId: @mediaItemId,
            mimeType: @mimeType
        } IN @@collectionName
         RETURN OLD
        `;

        const bindVars = {
            '@collectionName': 'term__VIEWS',
            id: termId,
            mediaItemId,
            mimeType,
        };

        const cursor = await this.database
            .query({
                query,
                bindVars,
            })
            .catch((reason) => {
                throw new InternalError(
                    `Failed to add audio for term via term query repository: ${reason}`
                );
            });

        await cursor.all();
    }

    async fetchById(
        id: AggregateId
    ): Promise<Maybe<ITermViewModel & { actions: ICommandFormAndLabels[] }>> {
        const result = await this.database.fetchById(id);

        if (isNotFound(result)) return result;

        const asView = mapDatabaseDocumentToAggregateDTO(result);

        return asView;
    }

    async fetchMany(): Promise<(ITermViewModel & { actions: ICommandFormAndLabels[] })[]> {
        const result = await this.database.fetchMany();

        return result.map(mapDatabaseDocumentToAggregateDTO);
    }

    async count(): Promise<number> {
        return this.database.getCount();
    }
}
