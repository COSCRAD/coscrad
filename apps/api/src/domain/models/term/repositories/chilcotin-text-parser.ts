import { ITextSymbol, TextSymbolType } from '@coscrad/api-interfaces';

export class ChilcotinTextParser {
    getSymbols(text: string): ITextSymbol[] {
        const chilcotinAlphabet = ['a', 'e', 'i', 'ɨ', 'u', 'o', 't', 'tl', 'y', 'g'];

        /**
         * TODO
         * 1. include support for manual exceptions list
         * 2. support parsing nasal vowels
         */
        if (text === 'naẑlhiny') {
            return ['n', 'a', 'ẑ', 'lh', 'in', 'y'].map((value) => ({
                value,
                type: TextSymbolType.letter,
            }));
        }

        const [last, symbols] = text.split('').reduce(
            ([current, allSymbols]: [string, ITextSymbol[]], nextLetter: string) => {
                if (chilcotinAlphabet.includes(current + nextLetter)) {
                    return [current + nextLetter, allSymbols];
                }

                if (chilcotinAlphabet.includes(current)) {
                    const symbolToAdd: ITextSymbol = {
                        value: current,
                        type: TextSymbolType.letter,
                    };

                    allSymbols.push(symbolToAdd);

                    // add the Chilcotin letter to the list and start again with the next letter as the candidate
                    return [nextLetter, allSymbols];
                }

                /**
                 * At this point, we have a letter and it isn't in the
                 * alphabet (even if you add the next char)
                 */
                const foreignSymbol: ITextSymbol = {
                    value: current,
                    type: TextSymbolType.foreignLetter,
                };

                allSymbols.push(foreignSymbol);

                return ['', allSymbols];
            },
            ['', []] as [string, ITextSymbol[]]
            // TODO fix types
        ) as [string, ITextSymbol[]];

        const result = symbols as ITextSymbol[];

        if (last.length > 0) {
            const lastSymbol: ITextSymbol = chilcotinAlphabet.includes(last)
                ? {
                      value: last,
                      type: TextSymbolType.letter,
                  }
                : {
                      value: last,
                      type: TextSymbolType.foreignLetter,
                  };

            result.push(lastSymbol);
        }

        return result;
    }
}
