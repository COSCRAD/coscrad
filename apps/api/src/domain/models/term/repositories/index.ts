export * from './arango-term-query-repository';
