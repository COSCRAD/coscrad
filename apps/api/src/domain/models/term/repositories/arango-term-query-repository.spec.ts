import {
    AggregateType,
    IDetailQueryResult,
    ITermViewModel,
    LanguageCode,
    MultilingualTextItemRole,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import buildMockConfigService from '../../../../app/config/__tests__/utilities/buildMockConfigService';
import buildConfigFilePath from '../../../../app/config/buildConfigFilePath';
import { Environment } from '../../../../app/config/constants/Environment';
import { InternalError } from '../../../../lib/errors/InternalError';
import { isNotFound, NotFound } from '../../../../lib/types/not-found';
import { ArangoConnectionProvider } from '../../../../persistence/database/arango-connection.provider';
import { ArangoDatabaseForCollection } from '../../../../persistence/database/arango-database-for-collection';
import { ArangoCollectionId } from '../../../../persistence/database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import mapDatabaseDocumentToAggregateDTO from '../../../../persistence/database/utilities/mapDatabaseDocumentToAggregateDTO';
import mapEntityDTOToDatabaseDocument from '../../../../persistence/database/utilities/mapEntityDTOToDatabaseDocument';
import { PersistenceModule } from '../../../../persistence/persistence.module';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import { ArangoRepositoryForAggregate } from '../../../../persistence/repositories/arango-repository-for-aggregate';
import { TermViewModel } from '../../../../queries/buildViewModelForResource/viewModels/term.view-model';
import { TestEventStream } from '../../../../test-data/events';
import getValidAggregateInstanceForTest from '../../../__tests__/utilities/getValidAggregateInstanceForTest';
import { buildMultilingualTextWithSingleItem } from '../../../common/build-multilingual-text-with-single-item';
import { MultilingualText, MultilingualTextItem } from '../../../common/entities/multilingual-text';
import buildInstanceFactory from '../../../factories/utilities/buildInstanceFactory';
import { IRepositoryForAggregate } from '../../../repositories/interfaces/repository-for-aggregate.interface';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import { AudioItemCreated } from '../../audio-visual/audio-item/commands/create-audio-item/audio-item-created.event';
import { EventSourcedAudioItemViewModel } from '../../audio-visual/audio-item/queries';
import { IAudioItemQueryRepository } from '../../audio-visual/audio-item/queries/audio-item-query-repository.interface';
import { ArangoAudioItemQueryRepository } from '../../audio-visual/audio-item/repositories/arango-audio-item-query-repository';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { CoscradContributor } from '../../user-management/contributor';
import { FullName } from '../../user-management/user/entities/user/full-name.entity';
import { ITermQueryRepository } from '../queries/term-query-repository.interface';
import { ArangoTermQueryRepository } from './arango-term-query-repository';

describe(`ArangoTermQueryRepository`, () => {
    let testQueryRepository: ITermQueryRepository;

    let audioItemQueryRepository: IAudioItemQueryRepository;

    let databaseProvider: ArangoDatabaseProvider;

    let arangoDatabaseForCollection: ArangoDatabaseForCollection<
        IDetailQueryResult<ITermViewModel>
    >;

    let contributorRepository: IRepositoryForAggregate<CoscradContributor>;

    let app: INestApplication;

    beforeAll(async () => {
        const moduleRef = await Test.createTestingModule({
            imports: [PersistenceModule.forRootAsync()],
        })
            .overrideProvider(ConfigService)
            .useValue(
                buildMockConfigService(
                    {
                        ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
                        // TODO this shouldn't be necessary
                        ARANGO_DB_HOST_PORT: 8551,
                    },
                    buildConfigFilePath(Environment.test)
                )
            )
            .compile();

        await moduleRef.init();

        app = moduleRef.createNestApplication();

        const connectionProvider = app.get(ArangoConnectionProvider);

        databaseProvider = new ArangoDatabaseProvider(connectionProvider);

        arangoDatabaseForCollection = databaseProvider.getDatabaseForCollection('term__VIEWS');

        // TODO Use the DI system so this is more extensible to keep test maintenance lower
        audioItemQueryRepository = new ArangoAudioItemQueryRepository(connectionProvider);

        testQueryRepository = new ArangoTermQueryRepository(
            connectionProvider,
            audioItemQueryRepository
        );

        /**
         * Currently, the contributors are snapshot based (not event sourced).
         */
        contributorRepository = new ArangoRepositoryForAggregate(
            databaseProvider,
            ArangoCollectionId.contributors,
            buildInstanceFactory(CoscradContributor),
            mapDatabaseDocumentToAggregateDTO,
            mapEntityDTOToDatabaseDocument
        );
    });

    afterAll(async () => {
        databaseProvider.close();
    });

    const termIds = [1, 2, 3].map(buildDummyUuid);

    const buildTermText = (id: string) => `term ${id}`;

    const termText = buildTermText(termIds[0]);

    const originalLanguageCode = LanguageCode.Chilcotin;

    const translationLangaugeCode = LanguageCode.English;

    const textTranslation = 'foobar';

    const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor);

    const contributorIds = [101, 102, 103].map(buildDummyUuid);

    const contributorIdsAndNames = contributorIds.map((contributorId) => ({
        contributorId,
        fullName: new FullName({
            firstName: `user`,
            lastName: contributorId,
        }),
    }));

    const testContributors = contributorIdsAndNames.map(({ contributorId, fullName }) =>
        dummyContributor.clone({
            id: contributorId,
            fullName,
        })
    );

    const termViews = termIds.map((id) => ({
        id,
        contributions: [],
        name: buildMultilingualTextWithSingleItem(buildTermText(id), originalLanguageCode),
        actions: [],
        isPublished: false,
        accessControlList: new AccessControlList(),
    }));

    describe(`fetchById`, () => {
        const targetTermId = termIds[0];

        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await testQueryRepository.create(termViews[0]);
        });

        describe(`when there is a term with the given ID`, () => {
            it(`should return the expected view`, async () => {
                const result = await testQueryRepository.fetchById(targetTermId);

                expect(result).not.toBe(NotFound);

                const { name } = result as IDetailQueryResult<ITermViewModel>;

                const foundOriginalTextForTerm = name.items.find(
                    ({ languageCode }) => languageCode === originalLanguageCode
                ).text;

                expect(foundOriginalTextForTerm).toBe(termText);
            });
        });

        describe(`when there is no term with the given ID`, () => {
            it(`should return not found`, async () => {
                const result = await testQueryRepository.fetchById('BOGUS_123');

                expect(result).toBe(NotFound);
            });
        });
    });

    describe(`fetchMany`, () => {
        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            for (const term of termViews) {
                await testQueryRepository.create(term);
            }
        });

        it(`should return the expected term views`, async () => {
            const result = await testQueryRepository.fetchMany();

            expect(result).toHaveLength(termViews.length);
        });
    });

    describe(`count`, () => {
        describe(`when there are term views in the database`, () => {
            beforeEach(async () => {
                await arangoDatabaseForCollection.clear();

                for (const term of termViews) {
                    await testQueryRepository.create(term);
                }
            });

            it(`should return the expected result`, async () => {
                const result = await testQueryRepository.count();

                expect(result).toBe(termViews.length);
            });
        });

        describe(`when the database collection is empty`, () => {
            beforeEach(async () => {
                await arangoDatabaseForCollection.clear();

                // no terms are added here
            });

            it(`should return 0`, async () => {
                const result = await testQueryRepository.count();

                expect(result).toBe(0);
            });
        });
    });

    describe(`translate`, () => {
        const targetTerm = termViews[0];

        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await testQueryRepository.create(targetTerm);
        });

        it(`should append the expected multilingual text item`, async () => {
            const targetTranslationRole = MultilingualTextItemRole.freeTranslation;

            await testQueryRepository.translate(targetTerm.id, {
                text: textTranslation,
                languageCode: translationLangaugeCode,
                role: targetTranslationRole,
            });

            const updatedTerm = await testQueryRepository.fetchById(targetTerm.id);

            if (isNotFound(updatedTerm)) {
                expect(updatedTerm).not.toBe(NotFound);

                throw new InternalError('test failed');
            }

            const updatedName = new MultilingualText(updatedTerm.name); // we want an instance (not a DTO) for the query methods

            const searchResultForTranslation = updatedName.getTranslation(translationLangaugeCode);

            expect(searchResultForTranslation).not.toBe(NotFound);

            const foundTranslation = searchResultForTranslation as MultilingualTextItem;

            expect(foundTranslation.text).toBe(textTranslation);

            expect(foundTranslation.role).toBe(targetTranslationRole);
        });
    });

    describe(`addAudio`, () => {
        const targetTerm = termViews[0];

        const audioItemId = buildDummyUuid(55);

        const mediaItemId = buildDummyUuid(56);

        const audioCreationEvent = new TestEventStream()
            .andThen<AudioItemCreated>({
                type: 'AUDIO_ITEM_CREATED',
                payload: {
                    mediaItemId,
                },
            })
            .as({
                type: AggregateType.audioItem,
                id: audioItemId,
            })[0] as AudioItemCreated;

        const targetAudioItemView =
            EventSourcedAudioItemViewModel.fromAudioItemCreated(audioCreationEvent);

        beforeEach(async () => {
            // clear existing term views
            await arangoDatabaseForCollection.clear();

            // clear existing audio item views
            databaseProvider.getDatabaseForCollection('audioItem__VIEWS').clear();

            await testQueryRepository.create(targetTerm);

            await audioItemQueryRepository.create(targetAudioItemView);
        });

        it(`should append the audio item`, async () => {
            await testQueryRepository.addAudio(targetTerm.id, originalLanguageCode, audioItemId);

            const updatedView = (await testQueryRepository.fetchById(
                targetTerm.id
            )) as IDetailQueryResult<ITermViewModel>;

            // TODO In the future, we should use multilingual audio for terms
            expect(updatedView.mediaItemId).toBe(mediaItemId);
        });
    });

    describe(`allowUser`, () => {
        const targetTerm = termViews[0];

        beforeEach(async () => {
            // clear existing term views
            await arangoDatabaseForCollection.clear();

            // clear existing audio item views

            await testQueryRepository.create(targetTerm);
        });

        it(`should add the user to the ACL`, async () => {
            const userId = buildDummyUuid(456);

            await testQueryRepository.allowUser(targetTerm.id, userId);

            const updatedView = (await testQueryRepository.fetchById(
                targetTerm.id
            )) as TermViewModel;

            const updatedAcl = new AccessControlList(updatedView.accessControlList);

            const canUser = updatedAcl.canUser(userId);

            expect(canUser).toBe(true);
        });
    });

    describe(`delete`, () => {
        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await testQueryRepository.createMany(termViews);
        });

        it(`should remove the given term`, async () => {
            const targetTermViewId = termIds[0];

            const expectedNumberOfTermsAfterDelete = termViews.length - 1;

            await testQueryRepository.delete(targetTermViewId);

            const actualNumberOfTerms = await testQueryRepository.count();

            expect(actualNumberOfTerms).toBe(expectedNumberOfTermsAfterDelete);
        });
    });

    describe(`count`, () => {
        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await testQueryRepository.createMany(termViews);
        });

        it(`should return the correct count`, async () => {
            const result = await testQueryRepository.count();

            expect(result).toBe(termViews.length);
        });
    });

    describe(`create`, () => {
        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();
        });

        it(`should create the correct term view`, async () => {
            const termToCreate = termViews[0];

            // act
            await testQueryRepository.create(termToCreate);

            const searchResult = await testQueryRepository.fetchById(termToCreate.id);

            expect(searchResult).not.toBe(NotFound);

            const foundTermView = searchResult as TermViewModel;

            const name = new MultilingualText(foundTermView.name);

            expect(name.getOriginalTextItem().text).toBe(termText);
        });
    });

    describe(`createMany`, () => {
        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();
        });

        it(`should create the expected term views`, async () => {
            // act
            await testQueryRepository.createMany(termViews);

            const actualCount = await testQueryRepository.count();

            expect(actualCount).toBe(termViews.length);
        });
    });

    describe('publish', () => {
        const targetTerm = termViews[0];

        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await testQueryRepository.create(targetTerm);
        });

        it(`should publish the given term`, async () => {
            await testQueryRepository.publish(targetTerm.id);

            const updatedView = (await testQueryRepository.fetchById(
                targetTerm.id
            )) as TermViewModel;

            expect(updatedView.isPublished).toBe(true);
        });
    });

    describe(`attribute`, () => {
        const targetTerm = termViews[0];

        beforeEach(async () => {
            await arangoDatabaseForCollection.clear();

            await databaseProvider.getDatabaseForCollection('contributors').clear();

            await testQueryRepository.create(targetTerm);

            await contributorRepository.createMany(testContributors);
        });

        it(`should add the given contributions`, async () => {
            await testQueryRepository.attribute(
                targetTerm.id,
                testContributors.map((c) => c.id)
            );

            const updatedView = (await testQueryRepository.fetchById(
                targetTerm.id
            )) as TermViewModel;

            const missingAttributions = updatedView.contributions.filter(
                ({ id }) => !contributorIds.includes(id)
            );

            expect(missingAttributions).toHaveLength(0);

            const { contributorId: targetContributorId, fullName: expectedFullName } =
                contributorIdsAndNames[0];

            const contributionForFirstUser = updatedView.contributions.find(
                ({ id }) => id === targetContributorId
            );

            expect(contributionForFirstUser.fullName).toBe(
                `${expectedFullName.firstName} ${expectedFullName.lastName}`
            );
        });
    });
});
