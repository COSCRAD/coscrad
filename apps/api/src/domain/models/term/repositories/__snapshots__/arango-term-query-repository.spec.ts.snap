// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArangoTermQueryRepository attribute when there are contributor IDs on the event meta should add the given contributions 1`] = `"English prompt created by: user 9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101, user 9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102, user 9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103"`;
